### PR TITLE
Refactoring of main.rs into modules and introduction of futures

### PR DESCRIPTION
**Pull request by vkomenda**
I did several different things in this PR:

1. Converted explicit threading code into futures and tested that it works.
2. Removed (possibly not all) calls to `unwrap`.
3. Introduced structured location-based errors using `failure`.
4. Introduced structured command-line options with `docopt`.
5. Updated [`README.md`](https://github.com/Fantom-foundation/full-cli-rs/blob/8d96374a3a2994e416a93ca8f25e204488bf0774/README.md) with instructions to run the binary.

There are TODOs and FIXMEs that I left in the code regarding what I think is still missing or incomplete (considering this initial template only, not the future functionality). Please feel free to insist that those are to be done in this PR. However, I must say that some of those TODOs are unclear to me and may require some reconnaissance.

**NOTE**: https://github.com/Fantom-foundation/libvm-rs/pull/1 is needed for the binary to be compiled by the latest nightly. Otherwise there will be a compilation error in `libvm-rs`.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -33,62 +58,11 @@ struct PeerConfig {
...
-    }
-}
-
 /// CLI environment obtained by interpreting the initial configuration.
+// FIXME: Requires clarification of what it is supposed to accomplish.
 struct Env {
-    // peers: Vec<TcpPeer>,
-    // node: Arc<TcpNode<Swirlds<TcpPeer, BTreeHashgraph>>>,
-    app: TcpApp,
+    // TODO: Why was this in the original code at all? It's not used so far.
```
Comment (AgustinCB) [inline src/main.rs]: It's used in line 136 of the original code.

File: src/server.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        let address: SocketAddr = format!("{}:{}", LOCALHOST, port)
+            .parse()
+            .map_err(Error::Parse)?;
+        let mut listener = TcpListener::bind(&address).map_err(Error::Bind)?;
+        let node_ref = Arc::new(self.node.clone());
+        loop {
+            let node_ref = node_ref.clone();
+            let (mut socket, _) = listener.accept().await.map_err(Error::Accept)?;
+            tokio::spawn(async move {
+                let mut buf: [u8; 4096] = [0; 4096];
```
Comment (AgustinCB) [inline src/server.rs]: Can we extract those magic values into proper constants?

**Review by AgustinCB — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 1 above)
Comment (vkomenda) [reply src/main.rs]: You are right. I accidentally omitted that line. Or maybe not accidentally: `TcpApp` would need to become a `Future` in order to be executed in parallel with `Server`.

File: src/server.rs (hunk 2 above)
Comment (vkomenda) [reply src/server.rs]: Sure, will do.

**Comment by vkomenda** [comment]
Now we need https://github.com/Fantom-foundation/libconsensus-lachesis-rs/pull/7 for the PR to compile.

**Comment by AgustinCB** [comment]
Don't forget to squash when merging.

**Review by AgustinCB — APPROVED** [review]

File: src/executor.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    /// The node that produces programs.
+    node: Arc<TcpNode<Swirlds<TcpPeer, BTreeHashgraph>>>,
+    /// The index of the next program.
+    program_index: usize,
+}
+
+impl Future for NewPrograms {
+    type Output = Vec<Program>;
+
+    fn poll(self: Pin<&mut Self>, _cxt: &mut Context) -> Poll<Self::Output> {
```
Comment (Maxime2) [inline src/executor.rs:27]: This is non-conventional implementation of the poll() method for Future trait, it misses to take waker out of Context and calling it when ready to be polled again.

**Pull request merged**